File: api/app/services/rss_service.py
from feedgen.feed import FeedGenerator
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import settings
from app.services.article_service import get_articles


async def generate_rss_feed(session: AsyncSession) -> str:
    """最新記事から RSS 2.0 XML を生成する。"""
    articles, _ = await get_articles(session, page=1, per_page=20)

    fg = FeedGenerator()
    fg.title("Tech News Curator")
    fg.link(href=settings.public_url, rel="alternate")
    fg.description("海外テックニュースの日本語要約")
    fg.language("ja")

    for article in articles:
        fe = fg.add_entry()
        fe.id(str(article.id))
        fe.title(article.title_ja or article.title_original or "Untitled")
        fe.link(href=article.source_url)
        if article.summary_ja:
            fe.description(article.summary_ja)
        if article.published_at:
            fe.pubDate(article.published_at)
        if article.author:
            fe.author(name=article.author)

    return fg.rss_str(pretty=True).decode("utf-8")

File: api/src/services/rss-service.ts
import { Feed } from "feed";
import { config } from "../config.js";
import type { DB } from "../database.js";
import { getArticles } from "./article-service.js";

export async function generateRssFeed(db: DB): Promise<string> {
  const { items: articleList } = await getArticles(db, 1, 20);

  const feed = new Feed({
    title: "Tech News Curator",
    description: "海外テックニュースの日本語要約",
    id: config.publicUrl,
    link: config.publicUrl,
    language: "ja",
    copyright: "",
  });

  for (const article of articleList) {
    feed.addItem({
      id: article.id,
      title: article.titleJa ?? article.titleOriginal ?? "Untitled",
      link: article.sourceUrl,
      description: article.summaryJa ?? undefined,
      date: article.publishedAt ?? article.createdAt,
      author: article.author ? [{ name: article.author }] : undefined,
    });
  }

  return feed.rss2();
}
